python/ray/experimental/sgd/pytorch/utils.py

```diff
@@ -45,42 +97,85 @@ def train(model, train_iterator, criterion, optimizer, config):
         with timers["grad"]:
             # compute gradients in a backward pass
             optimizer.zero_grad()
-            loss.backward()
+
+            if config.get(USE_FP16):
+                with amp.scale_loss(loss, optimizer) as scaled_loss:
+                    scaled_loss.backward()
+            else:
+                loss.backward()
 
         with timers["apply"]:
             # Call step of optimizer to update model params
             optimizer.step()
 
+        if scheduler and config.get(SCHEDULER_STEP) == SCHEDULER_STEP_BATCH:
+            scheduler.step()
+
         # measure elapsed time
         batch_time.update(time.time() - end)
         end = time.time()
 
+        if config.get(TEST_MODE) and batch_idx == 0:
+            break
+
+    if scheduler and config.get(SCHEDULER_STEP) == SCHEDULER_STEP_EPOCH:
+        scheduler.step()
+
     stats = {
         "batch_time": batch_time.avg,
-        "batch_processed": losses.count,
+        BATCH_COUNT: batch_idx + 1,
         "train_loss": losses.avg,
         "data_time": data_time.avg,
     }
     stats.update({k: t.mean for k, t in timers.items()})
     return stats
 
 
-def validate(model, val_iterator, criterion, config):
-    if isinstance(model, collections.Iterable):
+def validate(config, model, val_iterator, criterion, scheduler=None):
+    """Runs one standard validation pass over the val_iterator.
+
+    This function automatically measures timing for various operations such
+    as host to device transfer and processing time for the batch.
+
+    It also automatically detects and places the data on the given GPU device
+    if available.
+
+    Raises:
+        ValueError if multiple models/schedulers are provided. You
+            are expected to have a custom training function if you wish
+            to use multiple models/schedulers.
+
+    Args:
+        config: (dict): A user configuration provided into the Trainer
+            constructor.
+        model: The model as created by the model_creator.
+        train_iterator: An iterator created from the DataLoader which
+            wraps the provided Dataset.
+        criterion: The loss object created by the loss_creator.
+        scheduler (optional): The torch.optim.lr_scheduler object
+            as created by the scheduler_creator. By default,
+            this is not used in this function.
+
+    Returns:
+        A dict of metrics from the evaluation.
```

Comment:
Does the format of this dictionary matter?

Reply:
Nope